### PR TITLE
Add Storage

### PR DESCRIPTION
**Pull request by seowalex**
Add storage models, publishers and methods.

# How to use
* To save recipes/ingredients, use the `saveRecipe(_ recipe: inout Recipe)` and `saveIngredient(_ ingredient: inout Ingredient)` methods. Extend the `Recipe` and `Ingredient` structures for the in-app models.
* To load recipes/ingredients, subscribe to the appropriate [publisher](https://developer.apple.com/documentation/combine/receiving-and-handling-events-with-combine) (e.g. `recipesOrderedByNamePublisher()`) by attaching a [subscriber](https://developer.apple.com/documentation/combine/fail/sink(receivevalue:)).

# TODO
* [x] Let publisher publish associated records
* [x] Include stubs for translation of in-app models to database models and vice versa
* [x] Finish writing tests
* [x] Add categories to recipes/ingredients
* [ ] ~Add filters/sorting to publishers~

### CODE REVIEW

File: ChopChop/ChopChopApp.swift [hunk 1]
```diff
@@ -1,14 +1,9 @@
...
-//  ChopChop
-//
-//  Created by Chrystal on 9/3/21.
-//
-
 import SwiftUI
 
 @main
 struct ChopChopApp: App {
+    let appDatabase = AppDatabase.shared
```
Comment (chrystalquek) [inline ChopChop/ChopChopApp.swift:5]: I think for future only the model component should know about the database, should we put it somewhere In the storage manager?

File: ChopChop/Storage/AppDatabase+Persistence.swift [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import Foundation
+import GRDB
+
+extension AppDatabase {
+    static let shared = makeShared()
```
Comment (chrystalquek) [inline ChopChop/Storage/AppDatabase+Persistence.swift:5]: I think the graders don't really like Singletons, I had to remove my Singleton database thing for Peggle

File: ChopChop/Storage/AppDatabase.swift [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+                t.column("recipeId", .integer)
+                    .notNull()
+                    .indexed()
+                    .references("recipe", onDelete: .cascade)
+                t.column("name", .text)
+                    .notNull()
+                    .check { $0 != "" }
+                t.column("quantity", .text)
+                    .notNull()
+                t.uniqueKey(["recipeId", "name"])
```
Comment (chrystalquek) [inline ChopChop/Storage/AppDatabase.swift:53]: not sure if i'm interpreting this right but i think recipeIngredient's name does not need to be unique

File: ChopChop/Models/Ingredient.swift [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import Foundation
+import GRDB
+
+struct Ingredient: Equatable {
+    var id: Int64?
+    var ingredientCategoryId: Int64?
```
Comment (sc-arecrow) [inline ChopChop/Models/Ingredient.swift:6]: Is it possible to only have this property in `IngredientRecord`?

File: ChopChop/Models/Recipe.swift [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import GRDB
+
+struct Recipe: Equatable {
+    var id: Int64?
+    var recipeCategoryId: Int64?
+    var name: String
+    var ingredients: [String: Quantity]
```
Comment (sc-arecrow) [inline ChopChop/Models/Recipe.swift]: This will need to be updated to work with `IngredientReference` in the future

**Review by chrystalquek — APPROVED** [review]

File: ChopChop/Models/Ingredient.swift [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    var name: String
+    var sets: [Date?: Quantity]
```
Comment (sc-arecrow) [inline ChopChop/Models/Ingredient.swift]: Should we go ahead with this implementation (map of dates to quantities) or the initial plan of having a sorted array of `IngredientBatch`es?

**Review by sc-arecrow — COMMENTED** [review]
Perhaps an alternative name for `IngredientSet` could be `IngredientBatch`

File: ChopChop/Storage/AppDatabase.swift (hunk 3 above)
Comment (shadowezz) [reply ChopChop/Storage/AppDatabase.swift:53]: I think it means the entire pair is unique

**Review by shadowezz — APPROVED** [review]
LGTM other than the issues alr brought up

File: ChopChop/Models/Ingredient.swift (hunk 6 above)
Comment (seowalex) [reply ChopChop/Models/Ingredient.swift]: Similar to above, I just used the simplest data structure for now, subject to change.

File: ChopChop/Models/Recipe.swift (hunk 5 above)
Comment (seowalex) [reply ChopChop/Models/Recipe.swift]: For now I wasn't very sure what the rest of you were going to use for your models, so I just used the simplest data structure, which can of course be modified. We just need to remember to update the mapping from the Record to the actual model.

File: ChopChop/Models/Ingredient.swift (hunk 4 above)
Comment (seowalex) [reply ChopChop/Models/Ingredient.swift:6]: Sure. The properties for the in-model models can all be changed, as long as the mappings are updated from the records.

File: ChopChop/Storage/AppDatabase.swift (hunk 3 above)
Comment (seowalex) [reply ChopChop/Storage/AppDatabase.swift:53]: Yup this means it's unique on both the `recipeId` and `name` columns, which mean that each recipe cannot have ingredients with the same name.

File: ChopChop/Storage/AppDatabase+Persistence.swift (hunk 2 above)
Comment (seowalex) [reply ChopChop/Storage/AppDatabase+Persistence.swift:5]: Sure I'll revise this.

File: ChopChop/ChopChopApp.swift (hunk 1 above)
Comment (seowalex) [reply ChopChop/ChopChopApp.swift:5]: So you're suggesting that only the storage manager should access the database, and all classes should modify the database through the storage manager? Originally that was my thought as well, but I realised there would be quite a number of duplicated methods, where the storage manager simply calls the database manager's version of the method.

File: ChopChop/Storage/StorageManager.swift [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+    // MARK: - Storage Manager: Create/Update
+
+    func saveRecipe(_ recipe: inout Recipe) throws {
+        var recipeRecord = RecipeRecord(id: recipe.id, recipeCategoryId: recipe.recipeCategoryId, name: recipe.name)
+        var ingredientRecords = recipe.ingredients.map { name, quantity in
+            RecipeIngredientRecord(recipeId: recipe.id, name: name, quantity: quantity)
+        }
+        var stepRecords = recipe.steps.enumerated().map { index, content in
+            RecipeStepRecord(recipeId: recipe.id, index: index + 1, content: content)
```
Comment (chrystalquek) [inline ChopChop/Storage/StorageManager.swift:18]: Why +1

**Pull request merged**